diff --git a/Simulators/gem5/nvmain_mem.cc b/Simulators/gem5/nvmain_mem.cc
index e8ce704..5fa222b 100644
--- a/Simulators/gem5/nvmain_mem.cc
+++ b/Simulators/gem5/nvmain_mem.cc
@@ -241,11 +241,11 @@ void NVMainMemory::wakeup()
 }
 
 
-BaseSlavePort &
-NVMainMemory::getSlavePort(const std::string& if_name, PortID idx)
+Port &
+NVMainMemory::getPort(const std::string& if_name, PortID idx)
 {
     if (if_name != "port") {
-        return MemObject::getSlavePort(if_name, idx);
+        return AbstractMemory::getPort(if_name, idx);
     } else {
         return port;
     }
@@ -301,7 +301,7 @@ NVMainMemory::SetRequestData(NVMainRequest *request, PacketPtr pkt)
 
     if (pkt->isRead())
     {
-        Request *dataReq = new Request(pkt->getAddr(), pkt->getSize(), 0, Request::funcMasterId);
+	const RequestPtr dataReq = std::make_shared<Request>(pkt->getAddr(), pkt->getSize(), 0, Request::funcMasterId);
         Packet *dataPkt = new Packet(dataReq, MemCmd::ReadReq);
         dataPkt->allocate();
         doFunctionalAccess(dataPkt);
@@ -316,12 +316,11 @@ NVMainMemory::SetRequestData(NVMainRequest *request, PacketPtr pkt)
         }
 
         delete dataPkt;
-        delete dataReq;
         delete [] hostAddr;
     }
     else
     {
-        Request *dataReq = new Request(pkt->getAddr(), pkt->getSize(), 0, Request::funcMasterId);
+	const RequestPtr dataReq = std::make_shared<Request>(pkt->getAddr(), pkt->getSize(), 0, Request::funcMasterId);
         Packet *dataPkt = new Packet(dataReq, MemCmd::ReadReq);
         dataPkt->allocate();
         doFunctionalAccess(dataPkt);
@@ -339,7 +338,6 @@ NVMainMemory::SetRequestData(NVMainRequest *request, PacketPtr pkt)
         }
 
         delete dataPkt;
-        delete dataReq;
         delete [] hostAddrT;
         delete [] hostAddr;
     }
@@ -409,10 +407,6 @@ NVMainMemory::MemoryPort::recvFunctional(PacketPtr pkt)
 
     memory.doFunctionalAccess(pkt);
 
-    for( std::deque<PacketPtr>::iterator i = memory.responseQueue.begin();
-         i != memory.responseQueue.end(); ++i )
-        pkt->checkFunctional(*i);
-
     pkt->popLabel();
 }
 
diff --git a/Simulators/gem5/nvmain_mem.hh b/Simulators/gem5/nvmain_mem.hh
index da7c217..b043e93 100644
--- a/Simulators/gem5/nvmain_mem.hh
+++ b/Simulators/gem5/nvmain_mem.hh
@@ -173,8 +173,8 @@ class NVMainMemory : public AbstractMemory, public NVM::NVMObject
     NVMainMemory(const Params *p);
     virtual ~NVMainMemory();
 
-    BaseSlavePort& getSlavePort(const std::string& if_name,
-                                PortID idx = InvalidPortID);
+    Port& getPort(const std::string& if_name,
+                  PortID idx = InvalidPortID);
     void init();
     void startup();
     void wakeup();
